### PR TITLE
Power management and other comments + constify ipc_buffer_new

### PR DESCRIPTION
**Pull request by marc-hb**
See commit messages.

### CODE REVIEW

**Comment by marc-hb** [comment]
The only checkpatch warning is `WARNING: Consider removing the #if 1 and its #endif`

File: src/include/sof/audio/buffer.h
Comment (lgirdwood) [inline src/include/sof/audio/buffer.h:157]: We are probably missing other const usage wrt IPC too.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
> The only checkpatch warning is `WARNING: Consider removing the #if 1 and its #endif`

So CI is building and running successfully here with the new assert. I assume you have checked the binary size difference with XCC (assume GCC will be the same).

File: src/include/sof/audio/buffer.h
Comment (marc-hb) [reply src/include/sof/audio/buffer.h:157]: Probably, this one just came up when looking at your valgrind fixes.

**Review by marc-hb — COMMENTED** [review]
> So CI is building and running successfully here with the new assert. 

No, it's commented out.

> I assume you have checked the binary size difference with XCC (assume GCC will be the same).

I checked only with gcc10 and only with the default flags, that's why I left it commented out.

**Comment by lgirdwood** [comment]
> > So CI is building and running successfully here with the new assert.
> 
> No, it's commented out.
> 
> > I assume you have checked the binary size difference with XCC (assume GCC will be the same).
> 
> I checked only with gcc10 and only with the default flags, that's why I left it commented out.

ok, lets check with XCC before we merge this as we can remove the conditional or have an ```#if XCC``` instead.

**Comment by lgirdwood** [comment]
> > > So CI is building and running successfully here with the new assert.
> > 
> > 
> > No, it's commented out.
> > > I assume you have checked the binary size difference with XCC (assume GCC will be the same).
> > 
> > 
> > I checked only with gcc10 and only with the default flags, that's why I left it commented out.
> 
> ok, lets check with XCC before we merge this as we can remove the conditional or have an `#if XCC` instead.

oh - CI probably has the info for XCC builds ?

**Comment by marc-hb** [comment]
I tested gcc and xcc with various -O levels. I found only one case where the array-based assert increases the `.bss`: gcc with -O0. All other cases keep the sizes the same. The xcc versions I looked at optimize the unused array away even at the -O0 level.

Comment updated accordingly.

File: src/include/sof/common.h
Comment (lyakh) [inline src/include/sof/common.h:140]: does it have to be `static`? And there's an additional space there

Comment (marc-hb) [reply src/include/sof/common.h:140]: > does it have to be static?

Depends where you prefer to waste a few bytes per assertion at the `-O0` level. I suspect the .bss is likely to have less performance impact.

> And there's an additional space there

That was for aligning with the other one but not very useful, I can remove.

Comment (marc-hb) [reply src/include/sof/common.h:140]: > Depends where you prefer to waste a few bytes per assertion at the -O0 level. I suspect the .bss is likely to have less performance impact.

Actually: I didn't check whether `gcc -O0` allocates unused local variables or not.

**Comment by lgirdwood** [comment]
> I tested gcc and xcc with various -O levels. I found only one case where the array-based assert increases the `.bss`: gcc with -O0. All other cases keep the sizes the same. The xcc versions I looked at optimize the unused array away even at the -O0 level.
> 
> Comment updated accordingly.

Do we still need the ```#if 1``` if not we can make this the default check or would you be happy to keep this around for more validation ?

**Comment by marc-hb** [comment]
> Do we still need the #if 1 if not we can make this the default check or would you be happy to keep this around for more validation ?

I don't like `#if 1` either but this is really tricky situation.

On one hand I discovered that the `typedef`-based solution can silently succeed with variables, which is really bad for something called `STATIC_ASSERT`!

On the other hand the array based solution is 100% reliable BUT it can waste some memory with some compilers at the `-O0` level. I can't test future compilers. So that's why I'm submitting it commented out: if anyone ever has some STATIC_ASSERT doubts like I had, they can quickly and locally use the array-based one and find their mistake.

**Comment by lgirdwood** [comment]
CI showung a non booting DUT and a know alsabat issue on APL nocodec, both unrelated.

**Pull request merged**
